azurelinuxagent/common/protocol/wire.py

```diff
@@ -721,8 +720,9 @@ def update_goal_state(self, forced=False):
                 new_goal_state = GoalState.fetch_full_goal_state_if_incarnation_different_than(self, self._goal_state.incarnation)
 
             if new_goal_state is not None:
+                previous_goal_state_fetch_timestamp = self._goal_state.fetch_timestamp if self._goal_state else datetime.utcnow()
```

Comment:
"previous" doesn't quite fit with the naming of the other variables. At this point we have self._goal_state and new_goal_state, but we don't have a "previous" goal state.

i'd leave this function unchanged and simply do 
`
self.goal_state_flusher.flush(self._goal_state.fetch_timestamp)
`
below

Reply:
I initially did just that, but `self._goal_state` is `None` the first time we fetch a goal state.

Reply:
If self._goal_state is None then saving the current files on disk using datetime.utcnow() would be incorrect. 

You probably want to check at the timestamp on GoalState.*.xml. Using that timestamp would actually work for all the cases since we save the file when we fetch the goal state. If you prefer to be more explicit than relying in the timestamp you could extend the XML with an extra property for the timestamp.

Reply:
One problem with checking the timestamp on GoalState.*.xml file -- `os.path.getmtime` returns the epoch time, but in the *local* timezone, and it's a mess to figure out what the timezone is and how to convert to UTC across different python versions.

I wanted to avoid the second approach since extending the XML would add extra computation (currently, we only parse the XML in the beginning and work with the string representation which is finally saved to disk), but I think I will take it.

Reply:
I decided to go with the timestamp of GoalState.*.xml since the code was getting to complicated and open for bugs in the latter approach with modifying the XML. This is because we internally only parse the XML once, at the beginning, and then just work with the string representation of it in memory and I didn't want to 1) add a hardcoded XML node with the timestamp as a string before saving the file, nor 2) add the XML node to the XML and re-parse it just in order to store it as a text file.